File: core/designSystem/src/main/kotlin/com/danielyan/gopeak/designsystem/components/GoPeakTextField.kt
package com.danielyan.gopeak.designsystem.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.painter.Painter
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.tooling.preview.PreviewLightDark
import androidx.compose.ui.unit.dp
import com.danielyan.gopeak.designsystem.theme.GoPeakTheme

/**
 * GoPeak text field. Has both a [leadingIcon] (decorative) and a
 * [trailingIcon] slot — the trailing slot is a composable so it can host an interactive control
 * such as a password-reveal toggle.
 */
@Composable
fun GoPeakTextField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    label: String? = null,
    placeholder: String? = null,
    leadingIcon: Painter? = null,
    trailingIcon: (@Composable () -> Unit)? = null,
    visualTransformation: VisualTransformation = VisualTransformation.None,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
    singleLine: Boolean = true,
    enabled: Boolean = true,
) {
    val colors = GoPeakTheme.colors
    Column(modifier) {
        if (label != null) {
            BasicTextLabel(label)
            Spacer(Modifier.height(4.dp))
        }
        BasicTextField(
            value = value,
            onValueChange = onValueChange,
            modifier = Modifier.fillMaxWidth(),
            enabled = enabled,
            singleLine = singleLine,
            textStyle = GoPeakTheme.typography.body.merge(
                androidx.compose.ui.text.TextStyle(color = colors.onSurface),
            ),
            cursorBrush = SolidColor(colors.primary),
            visualTransformation = visualTransformation,
            keyboardOptions = keyboardOptions,
            decorationBox = { innerTextField ->
                Row(
                    modifier = Modifier
                        .clip(GoPeakTheme.shapes.small)
                        .background(colors.surfaceVariant)
                        .border(1.dp, colors.outline, GoPeakTheme.shapes.small)
                        .padding(horizontal = 16.dp, vertical = 14.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    if (leadingIcon != null) {
                        Icon(
                            painter = leadingIcon,
                            contentDescription = null,
                            tint = colors.onSurfaceVariant,
                            modifier = Modifier.size(16.dp),
                        )
                        Spacer(Modifier.width(12.dp))
                    }
                    Box(Modifier.weight(1f)) {
                        if (value.isEmpty() && placeholder != null) {
                            androidx.compose.foundation.text.BasicText(
                                text = placeholder,
                                style = GoPeakTheme.typography.body.merge(
                                    androidx.compose.ui.text.TextStyle(
                                        color = colors.onSurfaceVariant.copy(alpha = 0.5f),
                                    ),
                                ),
                            )
                        }
                        innerTextField()
                    }
                    if (trailingIcon != null) {
                        Spacer(Modifier.width(12.dp))
                        trailingIcon()
                    }
                }
            },
        )
    }
}

@Composable
private fun BasicTextLabel(text: String) {
    androidx.compose.foundation.text.BasicText(
        text = text.uppercase(),
        style = GoPeakTheme.typography.label.merge(
            androidx.compose.ui.text.TextStyle(color = GoPeakTheme.colors.onSurfaceVariant),
        ),
    )
}

@PreviewLightDark
@Composable
private fun GoPeakTextFieldPreview() {
    GoPeakTheme {
        Box(
            Modifier
                .background(GoPeakTheme.colors.background)
                .padding(24.dp),
        ) {
            GoPeakTextField(
                value = "",
                onValueChange = {},
                label = "Full name",
                placeholder = "John Doe",
            )
        }
    }
}

File: iosApp/DesignSystem/Sources/GoPeakDesignSystem/Components/GoPeakTextField.swift
import SwiftUI

/// GoPeak text field — the SwiftUI counterpart of Android's `GoPeakTextField`.
///
/// Has a decorative `leadingIcon` (SF Symbol) and a `trailingIcon` view slot, so the trailing
/// slot can host an interactive control such as a password-reveal toggle. Set `isSecure` to mask
/// input (the caller flips it from a trailing button to implement reveal).
public struct GoPeakTextField<Trailing: View>: View {

    @Binding private var text: String
    private let label: String?
    private let placeholder: String?
    private let leadingSymbol: String?
    private let isSecure: Bool
    private let trailing: () -> Trailing

    public init(
        text: Binding<String>,
        label: String? = nil,
        placeholder: String? = nil,
        leadingIcon: String? = nil,
        isSecure: Bool = false,
        @ViewBuilder trailingIcon: @escaping () -> Trailing = { EmptyView() }
    ) {
        self._text = text
        self.label = label
        self.placeholder = placeholder
        self.leadingSymbol = leadingIcon
        self.isSecure = isSecure
        self.trailing = trailingIcon
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            if let label {
                Text(label.uppercased())
                    .goPeakTextStyle(GoPeakTheme.typography.label)
                    .foregroundStyle(GoPeakTheme.colors.onSurfaceVariant)
            }

            HStack(spacing: 12) {
                if let leadingSymbol {
                    Image(systemName: leadingSymbol)
                        .font(.system(size: 16))
                        .foregroundStyle(GoPeakTheme.colors.onSurfaceVariant)
                }

                field
                    .font(GoPeakTheme.typography.body.font)
                    .foregroundStyle(GoPeakTheme.colors.onSurface)
                    .tint(GoPeakTheme.colors.primary)

                trailing()
            }
            .padding(.horizontal, 16)
            .padding(.vertical, 14)
            .background(
                GoPeakTheme.shapes.rounded(GoPeakTheme.shapes.small)
                    .fill(GoPeakTheme.colors.surfaceVariant)
            )
            .overlay(
                GoPeakTheme.shapes.rounded(GoPeakTheme.shapes.small)
                    .stroke(GoPeakTheme.colors.outline, lineWidth: 1)
            )
        }
    }

    @ViewBuilder
    private var field: some View {
        let prompt = placeholder.map {
            Text($0).foregroundStyle(GoPeakTheme.colors.onSurfaceVariant.opacity(0.5))
        }
        if isSecure {
            SecureField("", text: $text, prompt: prompt)
        } else {
            TextField("", text: $text, prompt: prompt)
        }
    }
}

// MARK: - Previews

private struct TextFieldGallery: View {
    @State private var name = ""
    @State private var password = ""
    @State private var revealed = false

    var body: some View {
        VStack(spacing: 16) {
            GoPeakTextField(
                text: $name,
                label: "Full name",
                placeholder: "John Doe",
                leadingIcon: "person"
            )
            GoPeakTextField(
                text: $password,
                label: "Password",
                placeholder: "••••••••",
                leadingIcon: "lock",
                isSecure: !revealed
            ) {
                Button { revealed.toggle() } label: {
                    Image(systemName: revealed ? "eye.slash" : "eye")
                        .font(.system(size: 16))
                        .foregroundStyle(GoPeakTheme.colors.onSurfaceVariant)
                }
            }
        }
        .padding(24)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(GoPeakTheme.colors.background)
    }
}

#Preview("TextField – Dark") {
    TextFieldGallery().preferredColorScheme(.dark)
}

#Preview("TextField – Light") {
    TextFieldGallery().preferredColorScheme(.light)
}
